### PR TITLE
Linker: remove form listener in unlayout

### PR DESCRIPTION
**Pull request by calebcordry**
When unlayout callback is fired on amp-analytics, it's linker should stop listening for form submissions.

### CODE REVIEW

File: extensions/amp-analytics/0.1/amp-analytics.js [hunk 1]
```diff
@@ -183,6 +186,8 @@ export class AmpAnalytics extends AMP.BaseElement {
       });
     }
 
+    this.linkerManager_.removeListeners();
```
Comment (lannka) [inline extensions/amp-analytics/0.1/amp-analytics.js]: this can lead NPE. move it to L186

File: extensions/amp-analytics/0.1/linker-manager.js [hunk 2]
```diff
@@ -117,6 +120,15 @@ export class LinkerManager {
         });
   }
 
+  /**
+   * Remove any listeners created to manage form submission.
+   */
+  removeListeners() {
```
Comment (lannka) [inline extensions/amp-analytics/0.1/linker-manager.js]: let's call it `dispose()` to pair up with `init()`

File: extensions/amp-analytics/0.1/amp-analytics.js (hunk 1 above)
Comment (calebcordry) [reply extensions/amp-analytics/0.1/amp-analytics.js]: Done.

File: extensions/amp-analytics/0.1/linker-manager.js (hunk 2 above)
Comment (calebcordry) [reply extensions/amp-analytics/0.1/linker-manager.js]: Done.

File: extensions/amp-analytics/0.1/test/test-amp-analytics.js [hunk 3]
```diff
@@ -1393,6 +1412,8 @@ describes.realWin('amp-analytics', {
       });
 
       sandbox.stub(viewer, 'isVisible').returns(false);
+      analytics.linkerManager_ = {dispose: () => {}};
```
Comment (lannka) [inline extensions/amp-analytics/0.1/test/test-amp-analytics.js]: this is not needed. 

this test should actually capture the bug

File: extensions/amp-analytics/0.1/test/test-amp-analytics.js [hunk 4]
```diff
@@ -199,6 +200,24 @@ describes.realWin('amp-analytics', {
         expect(linkerStub.calledOnce).to.be.true;
       });
     });
+
+    it('Removes linker listeners.', () => {
```
Comment (lannka) [inline extensions/amp-analytics/0.1/test/test-amp-analytics.js]: this test is not useful. the bug should have been caught in one of the tests below

File: extensions/amp-analytics/0.1/amp-analytics.js [hunk 5]
```diff
@@ -180,9 +183,11 @@ export class AmpAnalytics extends AMP.BaseElement {
       this.iniPromise_.then(() => {
         // Page was unloaded - free up owned resources.
         this.transport_.deleteIframeTransport();
+        this.linkerManager_.dispose();
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/amp-analytics.js]: should not forget `detachedCallback`.

Comment (calebcordry) [reply extensions/amp-analytics/0.1/amp-analytics.js]: Good idea. Added.

File: extensions/amp-analytics/0.1/amp-analytics.js [hunk 6]
```diff
@@ -165,6 +173,7 @@ export class AmpAnalytics extends AMP.BaseElement {
       this.iniPromise_.then(() => {
         this.transport_.maybeInitIframeTransport(
             this.getAmpDoc().win, this.element);
+        this.linkerManager_.enableFormSupport();
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/amp-analytics.js]: Looks like you fail to check `'linker-form'` experiment or wait for `allLinkerPromises_ ` here.

Comment (calebcordry) [reply extensions/amp-analytics/0.1/amp-analytics.js]: Good catch. Added experiment protection. The `allLinkersPromises_` part is actually leftover from when we were waiting to scan the page for forms, and is no longer necessary, so it was removed.

File: extensions/amp-analytics/0.1/amp-analytics.js [hunk 7]
```diff
@@ -153,6 +156,11 @@ export class AmpAnalytics extends AMP.BaseElement {
       this.analyticsGroup_.dispose();
       this.analyticsGroup_ = null;
     }
+
+    if (this.linkerManager_) {
+      this.linkerManager_.dispose();
```
Comment (lannka) [inline extensions/amp-analytics/0.1/amp-analytics.js]: set to null

Comment (calebcordry) [reply extensions/amp-analytics/0.1/amp-analytics.js]: Done.

**Pull request merged**